examples/mnist_denoising_autoencoder.py

```diff
@@ -0,0 +1,139 @@
+true signal. 
+
+Noise + Data ---> Denoising Autoencoder ---> Data
+
+Given a training dataset of corrupted data as input and 
+true signal as output, a denoising autoencoder can recover the 
+hidden structure to generate clean data.
+
+This example has modular design. The encoder, decoder and autoencoder
+are 3 models that share weights. For example, after training the 
+autoencoder, the encoder can be used to  generate latent vectors
+of input data for low-dim visualization like PCA or TSNE.
+'''
+
+import keras
+from keras.layers import Activation, Dense, Input, BatchNormalization
+from keras.layers import Conv2D, MaxPooling2D, Flatten
+from keras.layers import Reshape, Conv2DTranspose, UpSampling2D
+from keras.models import Model
+from keras import backend as K
+from keras.datasets import mnist
+import numpy as np
+import matplotlib.pyplot as plt
+
+# MNIST dataset
+(x_train, _), (x_test, _) = mnist.load_data()
+
+image_size = x_train.shape[1]
+x_train = np.reshape(x_train, [-1, image_size, image_size, 1])
+x_test = np.reshape(x_test, [-1, image_size, image_size, 1])
+x_train = x_train.astype('float32') / 255
+x_test = x_test.astype('float32') / 255
+
+# Generate corrupted MNIST images by adding noise with normal dist
+# centered at 0.5 and std=0.5
+noise = np.random.normal(loc=0.5, scale=0.5, size=x_train.shape)
+x_train_noisy = x_train + noise
+noise = np.random.normal(loc=0.5, scale=0.5, size=x_test.shape)
+x_test_noisy = x_test + noise
+
+x_train_noisy = np.clip(x_train_noisy, 0., 1.)
+x_test_noisy = np.clip(x_test_noisy, 0., 1.)
+
+# Network parameters
+input_shape = (image_size, image_size, 1)
+batch_size = 128
+kernel_size = 3
+pool_size = 2
+filters = 16
+latent_dim = 16
+
+# Build the Autoencoder Model
+# First build the Encoder Model
+inputs = Input(shape=input_shape, name='encoder_input')
+x = inputs
+# Stack of ReLU-Conv2D blocks
+# Notes: 
+# 1) Use Batch Normalization before ReLU on deep networks
+# 2) Use MaxPooling2D as alternative to strides>1 
+# - faster but not as good as strides>1
+for i in range(2):
+    x = Activation('relu')(x)
+    filters = filters * 2
+    x = Conv2D(filters=filters,
+               kernel_size=kernel_size,
+               strides=2,
+               padding='same')(x)
+
+# Shape info needed to build decoder model
+shape = K.int_shape(x)
+
+# Generate the latent vector
+x = Flatten()(x)
+latent = Dense(latent_dim, name='latent_vector')(x)
+
+# Instantiate Encoder Model
+encoder = Model(inputs, latent, name='encoder')
+encoder.summary()
+
+# Build the Decoder Model
+latent_inputs = Input(shape=(latent_dim,), name='decoder_input')
+x = Dense(shape[1] * shape[2] * shape[3])(latent_inputs)
+x = Reshape((shape[1], shape[2], shape[3]))(x)
+
+# Stack of ReLU-Transposed Conv2D blocks
+# Notes: 
+# 1) Use Batch Normalization before ReLU on deep networks
+# 2) Use UpSampling2D as alternative to strides>1 
+# - faster but not as good as strides>1
+for i in range(2):
+    x = Activation('relu')(x)
```

Comment:
Better to put the activation as the `activation` keyword of the layer below

Reply:
Fixed. Thanks.